RFC.md

```diff
@@ -1,12 +1,289 @@
 # Introducing a new multidimensional array indexing intrinsic
 
-# Intrinsic specification
+## The request for change: A new instruction, `multidim_array_index`.
 
-# Example
-## allocating something of that shape
-## indexing into something of that shape
-## loop analyses that uses this
+We propose the addition of a new instruction, called `multidim_array_index`. 
+This will allow us to represent array indexing into an array `A[d1][d2][..][dk]`
+as `multidim_array_index A d1, d2, d3, ... dk` instead of flattening
+the information into `gep A, d1 * n1 + d2 * n2 + ... + dk * nk`. The former
+unflattened representation is advantageous for analysis and optimisation. It also
+allows us to represent array indexing semantics of languages such as Fortran
+and Chapel, which differs from that of C based array indexing.
+	
+## Motivating the need for a multi-dimensional array indexing intrinsic
+There are primarily one of two views of array indexing involving
+multiple dimensions most languages take, which we discuss to motivate 
+the need for a multi-dimensional array index. This consideration impacts
+the kinds of analysis we can perform on the program. In Polly, we care about
+dependence analysis, so the examples here will focus on that particular problem.
 
-# The math, how to express every conceivable array indexing scheme known to man
+Let us consider an array indexing operation of the form:
+```cpp
+int ex1(int n, int m, B[n][m], int x1, int x2, int y1, int y2) {
+	assert(x1 != x2);
+	assert(y1 != y2);
+	B[x1][y1] = 1;
+	printf("%d", B[x2][y2]);
+	exit(0);
+}
+```
 
+One would like to infer that since the array indices _interpreted as tuples_
+`(x1, y1)` and `(x2, y2)` do not have the same value, due to the guarding asserts
+that `x1 != x2` and `y1 != y2`. The write `B[x1][y1] = 1` can 
+in no way interfere with the value of `B[x2][y2]`. Consquently,
+we can optimise the program into:
+
+
+```cpp
+int ex1_opt(int n, int m, B[n][m], int x1, int x2, int y1, int y2) {
+    // B[x1][y1] = 1 is omitted because the result 
+    // cannot be used: 
+    //  It is not used in the print and then the program exits
+	printf("%d", B[x2][y2]);
+	exit(0);
+}
+```
+
+However, alas, this is illegal, for the C language does not provide
+semantics that allow the final inference above. It is conceivable that
+`x1 != x2, y1 != y2`, but the indices do actually alias, since
+according to C semantics, the two indices alias if the _flattened
+representation of the indices alias_. Consider the parameter
+values:
+
+```
+n = m = 3
+x1 = 1, y1 = 0; B[x1][y1] = nx1+y1 = 3*1+0=3
+x2 = 0, y2 = 3; B[x2][y2] = nx2+y2 = 3*0+3=3
```

Comment:
Is `y2 = 3` even legal?  I'm not a language lawyer.

Reply:
Yes it is, because the C language reference defines the semantics of an index `A[x]` as `*(A + x)`, and since the full index expression is inbounds, the array indexing is considered inbounds.

Reply:
There recently was a discussion on the llvm-dev list about this: https://lists.llvm.org/pipermail/llvm-dev/2019-May/132587.html

First, being convinced that it is defined. I then asked Hal Finkel (vice-president of the C++ standard committee) how to read the standard. After looking into it he thinks it is, in fact, undefined behavior.

It only matters to some extend because there is some code out there using this behavior, and we would still need delinerization if it was written as `B[x*n + y]`.

Reply:
@Meinersbur: I see, I was not aware of this. Even in papers that mention this, I've always seen this as being defined. So, how do you suggest to best re-word this paragraph? Is it better to write the de-linearlized representation?

Reply:
I honestly do not know.